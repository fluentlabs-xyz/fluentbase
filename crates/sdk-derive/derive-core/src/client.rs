use crate::{args::RouterArgs, function_id::FunctionIDAttribute, mode::RouterMode, route::Route};
use convert_case::{Case, Casing};
use proc_macro2::TokenStream as TokenStream2;
use quote::{format_ident, quote, ToTokens};
use syn::{
    parse::{Parse, ParseStream},
    Attribute,
    Result,
    TraitItem,
    TraitItemFn,
};

pub struct ClientMethod {
    route: Route,
    mode: RouterMode,
}

impl ClientMethod {
    fn new(method: &TraitItemFn, mode: RouterMode) -> Result<Self> {
        let attrs = &method.attrs;

        let mut route = Route::try_from(method)?;

        for attr in attrs {
            if attr.path().is_ident("function_id") {
                route.function_id_attr = Some(attr.parse_args::<FunctionIDAttribute>()?);
            }
        }

        Ok(ClientMethod { route, mode })
    }

    fn generate_codecs(&self) -> TokenStream2 {
        self.route.generate_codec_impl(&self.mode)
    }

    fn generate_helpers(&self) -> TokenStream2 {
        let fn_name = &self.route.sig().ident;

        let param_names = self
            .route
            .args
            .iter()
            .map(|arg| &arg.ident)
            .collect::<Vec<_>>();
        let param_types = self
            .route
            .args
            .iter()
            .map(|arg| &arg.ty)
            .collect::<Vec<_>>();

        let return_types = &param_types;

        let fn_name_str = fn_name.to_string();
        let encode_name = format_ident!("encode_{}", fn_name_str);
        let decode_name = format_ident!("decode_{}", fn_name_str);

        let pascal_name = self.route.fn_name.to_case(Case::Pascal);
        let call_struct = format_ident!("{}Call", pascal_name);

        quote! {
            pub fn #encode_name(
                &self,
                #(#param_names: #param_types,)*
            ) -> fluentbase_sdk::Bytes {
                 #call_struct::new((#(#param_names,)*)).encode().into()
            }

            pub fn #decode_name(
                &self,
                output: fluentbase_sdk::Bytes
            ) -> (#(#return_types,)*) {
                #call_struct::decode(&output)
                    .expect("failed to decode result")
                    .0
            }
        }
    }

    fn generate_implementation(&self) -> TokenStream2 {
        let fn_name = &self.route.sig().ident;
        let param_names = self
            .route
            .args
            .iter()
            .map(|arg| &arg.ident)
            .collect::<Vec<_>>();

        let param_types = self
            .route
            .args
            .iter()
            .map(|arg| &arg.ty)
            .collect::<Vec<_>>();
        let return_types = &param_types;

        let fn_name_str = fn_name.to_string();
        let encode_name = format_ident!("encode_{}", fn_name_str);
        let decode_name = format_ident!("decode_{}", fn_name_str);

        quote! {
            pub fn #fn_name(
                &mut self,
                contract_address: fluentbase_sdk::Address,
                value: fluentbase_sdk::U256,
                gas_limit: u64,
                #(#param_names: #param_types,)*
            ) -> (#(#return_types,)*) {
                use fluentbase_sdk::{TxContextReader, SyscallResult};

                let input = self.#encode_name(#(#param_names,)*);

                {
                    let context = self.sdk.context();
                    if context.tx_value() < value {
                        ::core::panic!("client: insufficient funds");
                    }
                    if context.tx_gas_limit() < gas_limit {
                        ::core::panic!("client: insufficient gas");
                    }
                }

                let result = self.sdk.call(
                    contract_address,
                    value,
                    &input,
                    Some(gas_limit),
                );

                if !SyscallResult::is_ok(result.status) {
                    ::core::panic!("client: call failed");
                }

                self.#decode_name(result.data)
            }
        }
    }
}

pub struct ClientGenerator {
    pub args: RouterArgs,
    trait_ast: syn::ItemTrait,
}

impl Parse for ClientGenerator {
    fn parse(input: ParseStream) -> Result<Self> {
        let attrs = input.call(Attribute::parse_outer)?;
        let args = if let Some(attr) = attrs.iter().find(|a| a.path().is_ident("client")) {
            attr.parse_args::<RouterArgs>()?
        } else {
            RouterArgs::new(RouterMode::default())
        };

        let trait_ast = input.parse()?;
        Ok(ClientGenerator { args, trait_ast })
    }
}

impl ClientGenerator {
    fn generate_client(&self) -> TokenStream2 {
        let trait_name = &self.trait_ast.ident;
        let client_name = format_ident!("{}Client", trait_name);

        let methods = self
            .trait_ast
            .items
            .iter()
            .filter_map(|item| {
                if let TraitItem::Fn(method) = item {
                    Some(ClientMethod::new(method, self.args.mode()))
                } else {
                    None
                }
            })
            .collect::<Result<Vec<_>>>()
            .unwrap_or_default();

        let codecs = methods.iter().map(|method| method.generate_codecs());
        let helpers = methods.iter().map(|method| method.generate_helpers());
        let implementations = methods
            .iter()
            .map(|method| method.generate_implementation());

        quote! {
            // Client structure
            pub struct #client_name<SDK> {
                pub sdk: SDK,
            }

            // Codec implementations
            #(#codecs)*

            // Helper functions
            impl<SDK: fluentbase_sdk::SharedAPI> #client_name<SDK> {
                pub fn new(sdk: SDK) -> Self {
                    Self { sdk }
                }

                #(#helpers)*
            }

            // Main interface implementation
            impl<SDK: fluentbase_sdk::SharedAPI> #client_name<SDK> {
                #(#implementations)*
            }
        }
    }
}

impl ToTokens for ClientGenerator {
    fn to_tokens(&self, tokens: &mut TokenStream2) {
        let client_impl = self.generate_client();
        tokens.extend(client_impl);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use prettyplease;
    use syn::parse_quote;

    #[test]
    fn test_generate_client() {
        let trait_ast: syn::ItemTrait = parse_quote! {
            pub trait TestContract {
                fn first_method(&self, value: u32) -> u32;
                fn second_method(&self, a: String, b: bool) -> (String, bool);
            }
        };

        let args = RouterArgs::new(RouterMode::default());
        let generator = ClientGenerator { args, trait_ast };

        let generated_tokens = generator.generate_client();
        let file = syn::parse_file(&generated_tokens.to_string()).unwrap();

        let formatted_code = prettyplease::unparse(&file);

        assert_snapshot!("generate_client", formatted_code);
    }
}
