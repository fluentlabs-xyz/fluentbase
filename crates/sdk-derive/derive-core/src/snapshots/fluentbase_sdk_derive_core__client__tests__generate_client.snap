---
source: crates/sdk-derive/derive-core/src/client.rs
expression: formatted_code
---
pub struct TestContractClient<SDK> {
    pub sdk: SDK,
}
pub type FirstMethodCallArgs = (u32,);
pub struct FirstMethodCall(FirstMethodCallArgs);
impl FirstMethodCall {
    pub const SELECTOR: [u8; 4] = [90u8, 193u8, 6u8, 45u8];
    pub const SIGNATURE: &'static str = "firstMethod(uint32)";
    pub fn new(args: FirstMethodCallArgs) -> Self {
        Self(args)
    }
    pub fn encode(&self) -> fluentbase_sdk::codec::bytes::Bytes {
        let mut buf = fluentbase_sdk::codec::bytes::BytesMut::new();
        fluentbase_sdk::codec::encoder::SolidityABI::encode(
                &(self.0.clone().0,),
                &mut buf,
                0,
            )
            .unwrap();
        let encoded_args = buf.freeze();
        let clean_args = if fluentbase_sdk::codec::encoder::SolidityABI::<
            FirstMethodCallArgs,
        >::is_dynamic() {
            encoded_args[32..].to_vec()
        } else {
            encoded_args.to_vec()
        };
        Self::SELECTOR.iter().copied().chain(clean_args).collect()
    }
    pub fn decode(
        buf: &impl fluentbase_sdk::codec::bytes::Buf,
    ) -> ::core::result::Result<Self, fluentbase_sdk::codec::CodecError> {
        use fluentbase_sdk::codec::bytes::BufMut;
        let dynamic_offset = if fluentbase_sdk::codec::encoder::SolidityABI::<
            FirstMethodCallArgs,
        >::is_dynamic() {
            ::fluentbase_sdk::U256::from(32).to_be_bytes::<32>().to_vec()
        } else {
            ::alloc::vec::Vec::new()
        };
        let mut combined_buf = fluentbase_sdk::codec::bytes::BytesMut::new();
        combined_buf.put_slice(&dynamic_offset);
        combined_buf.put_slice(buf.chunk());
        let args = fluentbase_sdk::codec::encoder::SolidityABI::<
            FirstMethodCallArgs,
        >::decode(&combined_buf.freeze(), 0)?;
        Ok(Self(args))
    }
}
impl ::core::ops::Deref for FirstMethodCall {
    type Target = FirstMethodCallArgs;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
pub type FirstMethodCallTarget = <FirstMethodCall as ::core::ops::Deref>::Target;
pub type FirstMethodReturnArgs = (u32,);
#[derive(Debug)]
pub struct FirstMethodReturn(FirstMethodReturnArgs);
impl FirstMethodReturn {
    pub fn new(args: FirstMethodReturnArgs) -> Self {
        Self(args)
    }
    pub fn encode(&self) -> fluentbase_sdk::codec::bytes::Bytes {
        let mut buf = fluentbase_sdk::codec::bytes::BytesMut::new();
        fluentbase_sdk::codec::encoder::SolidityABI::encode(
                &(self.0.clone().0,),
                &mut buf,
                0,
            )
            .unwrap();
        let encoded_args = buf.freeze();
        let clean_args = if fluentbase_sdk::codec::encoder::SolidityABI::<
            FirstMethodReturnArgs,
        >::is_dynamic() {
            encoded_args[32..].to_vec()
        } else {
            encoded_args.to_vec()
        };
        clean_args.into()
    }
    pub fn decode(
        buf: &impl fluentbase_sdk::codec::bytes::Buf,
    ) -> ::core::result::Result<Self, fluentbase_sdk::codec::CodecError> {
        use fluentbase_sdk::codec::bytes::BufMut;
        let dynamic_offset = if fluentbase_sdk::codec::encoder::SolidityABI::<
            FirstMethodReturnArgs,
        >::is_dynamic() {
            ::fluentbase_sdk::U256::from(32).to_be_bytes::<32>().to_vec()
        } else {
            ::alloc::vec::Vec::new()
        };
        let mut combined_buf = fluentbase_sdk::codec::bytes::BytesMut::new();
        combined_buf.put_slice(&dynamic_offset);
        combined_buf.put_slice(buf.chunk());
        let args = fluentbase_sdk::codec::encoder::SolidityABI::<
            FirstMethodReturnArgs,
        >::decode(&combined_buf.freeze(), 0)?;
        Ok(Self(args))
    }
}
impl ::core::ops::Deref for FirstMethodReturn {
    type Target = FirstMethodReturnArgs;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
pub type FirstMethodReturnTarget = <FirstMethodReturn as ::core::ops::Deref>::Target;
pub type SecondMethodCallArgs = (String, bool);
pub struct SecondMethodCall(SecondMethodCallArgs);
impl SecondMethodCall {
    pub const SELECTOR: [u8; 4] = [92u8, 23u8, 35u8, 57u8];
    pub const SIGNATURE: &'static str = "secondMethod(string,bool)";
    pub fn new(args: SecondMethodCallArgs) -> Self {
        Self(args)
    }
    pub fn encode(&self) -> fluentbase_sdk::codec::bytes::Bytes {
        let mut buf = fluentbase_sdk::codec::bytes::BytesMut::new();
        fluentbase_sdk::codec::encoder::SolidityABI::encode(
                &(self.0.clone().0, self.0.clone().1),
                &mut buf,
                0,
            )
            .unwrap();
        let encoded_args = buf.freeze();
        let clean_args = if fluentbase_sdk::codec::encoder::SolidityABI::<
            SecondMethodCallArgs,
        >::is_dynamic() {
            encoded_args[32..].to_vec()
        } else {
            encoded_args.to_vec()
        };
        Self::SELECTOR.iter().copied().chain(clean_args).collect()
    }
    pub fn decode(
        buf: &impl fluentbase_sdk::codec::bytes::Buf,
    ) -> ::core::result::Result<Self, fluentbase_sdk::codec::CodecError> {
        use fluentbase_sdk::codec::bytes::BufMut;
        let dynamic_offset = if fluentbase_sdk::codec::encoder::SolidityABI::<
            SecondMethodCallArgs,
        >::is_dynamic() {
            ::fluentbase_sdk::U256::from(32).to_be_bytes::<32>().to_vec()
        } else {
            ::alloc::vec::Vec::new()
        };
        let mut combined_buf = fluentbase_sdk::codec::bytes::BytesMut::new();
        combined_buf.put_slice(&dynamic_offset);
        combined_buf.put_slice(buf.chunk());
        let args = fluentbase_sdk::codec::encoder::SolidityABI::<
            SecondMethodCallArgs,
        >::decode(&combined_buf.freeze(), 0)?;
        Ok(Self(args))
    }
}
impl ::core::ops::Deref for SecondMethodCall {
    type Target = SecondMethodCallArgs;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
pub type SecondMethodCallTarget = <SecondMethodCall as ::core::ops::Deref>::Target;
pub type SecondMethodReturnArgs = (String, bool);
#[derive(Debug)]
pub struct SecondMethodReturn(SecondMethodReturnArgs);
impl SecondMethodReturn {
    pub fn new(args: SecondMethodReturnArgs) -> Self {
        Self(args)
    }
    pub fn encode(&self) -> fluentbase_sdk::codec::bytes::Bytes {
        let mut buf = fluentbase_sdk::codec::bytes::BytesMut::new();
        fluentbase_sdk::codec::encoder::SolidityABI::encode(
                &(self.0.clone().0, self.0.clone().1),
                &mut buf,
                0,
            )
            .unwrap();
        let encoded_args = buf.freeze();
        let clean_args = if fluentbase_sdk::codec::encoder::SolidityABI::<
            SecondMethodReturnArgs,
        >::is_dynamic() {
            encoded_args[32..].to_vec()
        } else {
            encoded_args.to_vec()
        };
        clean_args.into()
    }
    pub fn decode(
        buf: &impl fluentbase_sdk::codec::bytes::Buf,
    ) -> ::core::result::Result<Self, fluentbase_sdk::codec::CodecError> {
        use fluentbase_sdk::codec::bytes::BufMut;
        let dynamic_offset = if fluentbase_sdk::codec::encoder::SolidityABI::<
            SecondMethodReturnArgs,
        >::is_dynamic() {
            ::fluentbase_sdk::U256::from(32).to_be_bytes::<32>().to_vec()
        } else {
            ::alloc::vec::Vec::new()
        };
        let mut combined_buf = fluentbase_sdk::codec::bytes::BytesMut::new();
        combined_buf.put_slice(&dynamic_offset);
        combined_buf.put_slice(buf.chunk());
        let args = fluentbase_sdk::codec::encoder::SolidityABI::<
            SecondMethodReturnArgs,
        >::decode(&combined_buf.freeze(), 0)?;
        Ok(Self(args))
    }
}
impl ::core::ops::Deref for SecondMethodReturn {
    type Target = SecondMethodReturnArgs;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
pub type SecondMethodReturnTarget = <SecondMethodReturn as ::core::ops::Deref>::Target;
impl<SDK: fluentbase_sdk::SharedAPI> TestContractClient<SDK> {
    pub fn new(sdk: SDK) -> Self {
        Self { sdk }
    }
    pub fn encode_first_method(&self, value: u32) -> fluentbase_sdk::Bytes {
        FirstMethodCall::new((value,)).encode().into()
    }
    pub fn decode_first_method(&self, output: fluentbase_sdk::Bytes) -> (u32,) {
        FirstMethodCall::decode(&output).expect("failed to decode result").0
    }
    pub fn encode_second_method(&self, a: String, b: bool) -> fluentbase_sdk::Bytes {
        SecondMethodCall::new((a, b)).encode().into()
    }
    pub fn decode_second_method(&self, output: fluentbase_sdk::Bytes) -> (String, bool) {
        SecondMethodCall::decode(&output).expect("failed to decode result").0
    }
}
impl<SDK: fluentbase_sdk::SharedAPI> TestContractClient<SDK> {
    pub fn first_method(
        &mut self,
        contract_address: fluentbase_sdk::Address,
        value: fluentbase_sdk::U256,
        gas_limit: u64,
        value: u32,
    ) -> (u32,) {
        use fluentbase_sdk::{TxContextReader, SyscallResult};
        let input = self.encode_first_method(value);
        {
            let context = self.sdk.context();
            if context.tx_value() < value {
                ::core::panic!("client: insufficient funds");
            }
            if context.tx_gas_limit() < gas_limit {
                ::core::panic!("client: insufficient gas");
            }
        }
        let result = self.sdk.call(contract_address, value, &input, Some(gas_limit));
        if !SyscallResult::is_ok(result.status) {
            ::core::panic!("client: call failed");
        }
        self.decode_first_method(result.data)
    }
    pub fn second_method(
        &mut self,
        contract_address: fluentbase_sdk::Address,
        value: fluentbase_sdk::U256,
        gas_limit: u64,
        a: String,
        b: bool,
    ) -> (String, bool) {
        use fluentbase_sdk::{TxContextReader, SyscallResult};
        let input = self.encode_second_method(a, b);
        {
            let context = self.sdk.context();
            if context.tx_value() < value {
                ::core::panic!("client: insufficient funds");
            }
            if context.tx_gas_limit() < gas_limit {
                ::core::panic!("client: insufficient gas");
            }
        }
        let result = self.sdk.call(contract_address, value, &input, Some(gas_limit));
        if !SyscallResult::is_ok(result.status) {
            ::core::panic!("client: call failed");
        }
        self.decode_second_method(result.data)
    }
}
